app/services/notification_service.py

```
@@ -162,6 +359,28 @@ async def _dispatch_to_single_notifier(
         )
 
+        # Apply rate limiting for event-based notifications
+        # Only rate limit if we have metadata (indicates event-based dispatch)
+        if metadata:
+            event_type = metadata.get("event_type")
+            if event_type:
+                try:
+                    event_enum = NotificationEventType(event_type)
+                    if not rate_limiter.should_notify(notifier.id, event_enum):
+                        logger.info(
+                            f"Rate limited: Skipping notification to '{notifier.name}' for event {event_type.value}"
```

Ruff (0.14.13)

Logging statement uses f-string

(G004)